src/net.h

```diff
@@ -1257,6 +1276,40 @@ class CConnman
      */
     std::vector<CService> m_onion_binds;
 
+    /**
+     * RAII helper to atomically create a copy of `vNodes` and add a reference
+     * to each of the nodes. The nodes are released when this object is destroyed.
+     */
+    class NodesSnapshot
+    {
+    public:
+        explicit NodesSnapshot(const CConnman& connman) : m_connman{connman}
+        {
+            LOCK(m_connman.cs_vNodes);
+            m_nodes_copy = m_connman.vNodes;
+            for (auto& node : m_nodes_copy) {
+                node->AddRef();
+            }
+        }
+
+        ~NodesSnapshot()
+        {
+            LOCK(m_connman.cs_vNodes);
+            for (auto& node : m_nodes_copy) {
+                node->Release();
+            }
+        }
+
+        const std::vector<CNode*>& Nodes() const
```

Comment:
What do you think about just wrapping the m_nodes_copy `begin()` and `end()` functions:

```diff
diff --git a/src/net.cpp b/src/net.cpp
index 6040ee86be..5388af02e3 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -116,6 +116,10 @@ std::string strSubVersion;
  */
 class NodesSnapshot
 {
+    private:
+        const CConnman& m_connman;
+        const std::vector<CNode*> m_nodes_copy;
+
     public:
         explicit NodesSnapshot(const CConnman& connman) : m_connman{connman}, m_nodes_copy{m_connman.vNodes}
         {
@@ -133,14 +137,18 @@ class NodesSnapshot
             }
         }
 
-        const std::vector<CNode*>& Nodes() const
+        using iterator = decltype(m_nodes_copy.begin());
+
+        iterator begin()
         {
-            return m_nodes_copy;
+            return m_nodes_copy.begin();
+        }
+
+        iterator end()
+        {
+            return m_nodes_copy.end();
         }
 
-    private:
-        const CConnman& m_connman;
-        const std::vector<CNode*> m_nodes_copy;
 };
```

and then the client code can use a `NodeSnapshot` in range-based loops, eg:

```diff
→ git d
diff --git a/src/net.cpp b/src/net.cpp
index 6040ee86be..5388af02e3 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -1501,7 +1509,7 @@ void CConnman::SocketHandler()
     //
     // Service each socket
     //
-    for (CNode* pnode : snap.Nodes()) {
+    for (CNode* pnode : snap) {
         if (interruptNet)
             return;
 
```

Reply:
I think both ways would work.

The begin/end variant seems to be a bit more code - it saves just `.Nodes()` from the callers but has two methods and is also less friendly to grepping and code navigation - now I can point my "IDE" to `Nodes()` on the line `for (CNode* pnode : snap.Nodes()) {` and see what it returns or its definition. Not so with the `for (CNode* pnode : snap) {` line.

That said, I am fine either way but have a slight preference for the `.Nodes()` variant.

Reply:
No strong preference. Just pointing out an alternative. I'll mark this as resolved.